### DISCUSSION TITLE
A guide on manual verifications

### DISCUSSION THREAD

**Discussion by defido** [pcaversaccio/hardhat-project-template-ts#17] [Show and tell]
Still in testing mode...... Fantom scan presenting real issues in general verifying with constructor inputs. Unrelated to the starter. The below is just an idea/guide.

Just an idea to add an empty 'flatten' folder, then forge can flatten the contract for manual verifications. It seems forge strips the `//SPDX-License-Identifier:` automatically. As verification via the API is often sketchy, people knowing this idea, removes the need to search. Or it can be left as an 'IDEA' and people can see how I fixed my verification issues, and it might help them too.

This did multiple greatly positive things for me:

1) Fixed my verification issue within minutes doing a manual upload
2) Left the flattened file in a folder for other chains, as it works on all chains!
3) Forge automatically, and very quickly, removed all the `//SPDX-License-Identifier:` identifiers

First make sure you have done your compiling using [xDeployer](https://github.com/pcaversaccio/xdeployer) to the on chain factory [create2deployer](https://github.com/pcaversaccio/create2deployer)
Then:

`forge flatten --output flatten/Greeter.flattened.sol contracts/Greeter.sol`
`forge flatten --output <OUTPUT_DIRECTORY_FILE_NAME> <CONTRACT_LOCATION_CONTRACT_NAME.sol>`

You might think of a better folder output? Without a folder it just spurts the file into the main folder.

<img width="1016" alt="Screen Shot 2022-04-16 at 1 19 44 pm" src="https://user-images.githubusercontent.com/95591037/163659529-25ab5e44-49d7-40c4-8b95-8faaa78c3dbb.png">

Fill out the exact same parameters used in your projects version, license, choose 'single file', remember to turn on optimisation if you do optimisation runs, you can find this in the `hardhat.config.ts` file. Input your constructor arguments, often these are picked up automatically, find the ABI in the `abis` folder and match it to your contract name, in this case `Greeter.sol`.

```
cast abi-encode "constructor(string)" 'Hello, Hardhat!' <---- Seems to work, but won't verify, working on it.

```

<img width="1095" alt="Screen Shot 2022-04-16 at 1 32 27 pm" src="https://user-images.githubusercontent.com/95591037/163659993-d1a23f71-a674-46f6-86e5-36c57930420e.png">
<img width="1220" alt="Screen Shot 2022-04-16 at 1 32 54 pm" src="https://user-images.githubusercontent.com/95591037/163660016-7a08ce1e-5c95-4498-87d7-3b9426b74084.png">
<img width="1206" alt="Screen Shot 2022-04-16 at 1 38 38 pm" src="https://user-images.githubusercontent.com/95591037/163660109-5c55ab5b-d5e7-4f4b-a9f9-3e667ca2203a.png">

```
#!/usr/bin/env bash

## Read a config or .env for variables, use them in the script to auto run forge 

set -o allexport; source scripts/.verify.config; set +o allexport 
set -o allexport; source .env; set +o allexport 

encoded=`cast abi-encode $CONSTRUCTOR_STRING $ARGS`
echo ABI-encoded constructor arguments: ${encoded:2}

## The below is the encoded version of the argument  = here
0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000046865726500000000000000000000000000000000000000000000000000000000

## Prints our the arguments, not yet looping the arguments, ',' and spaces break this, even the current scripts break when using `Hello, Hardhat!` prints `Hello,`.

```
<img width="477" alt="Screen Shot 2022-04-16 at 8 29 57 pm" src="https://user-images.githubusercontent.com/95591037/163671567-de72ed3d-0438-4f3a-b22e-da9adc7a13b4.png">

**Comment by pcaversaccio** [comment]
Thanks for sharing your input @defido! My recommendation is usually the same in case the API-based verification fails: try using the flattened contract file.

As mentioned correctly by you, currently the [`flatten.sh`](https://github.com/pcaversaccio/hardhat-project-template-ts/blob/main/scripts/flatten.sh) generates a file in the root folder of the project. The idea of this template is that everyone can adjust the structure accordingly for their use case while the main scripts are already available. Verification can be a black box sometimes and it's kind of a centralisation issue since you can't debug the errors on the Etherscan/Blockscout side.

**Reply by pcaversaccio** [reply]
Re the compiler versions, you can use the list here: https://etherscan.io/solcversions

**Reply by defido** [reply]
I'll leave the file output as root then and make this a guide, as it's common, especially with Fantom & Polygon to fail verification + blockscout or whatever else people need..

The above will be edited a bit cause I am still testing, was able to read env or config via bash, and do auto print outs. But probably not something you want to add. And Forge verify seems, a little touchy/beta for now. Although, think forge as a long term use has real merit.

`chmod +x` <-- Possibly needed to run first to read permissions
`set -o allexport; source .env; set +o allexport` <--- Read env
`set -o allexport; source scripts/.verify.config; set +o allexport` <--- Read a config

Then use `$ENV_NAME` to use them

I setup a config file as I wanted to run tests over and over, doing it line by line was just too slow for that use, but in general line by line is ok for one offs or making a seperate bash to the same doesn't take long.

<img width="537" alt="Screen Shot 2022-04-16 at 8 28 32 pm" src="https://user-images.githubusercontent.com/95591037/163671716-ba9c6633-f44c-4abc-a535-642dd4446754.png">

**Reply by pcaversaccio** [reply]
Thx for debugging and sharing this information with everyone else! I let Forge evolve for some time until it's more stable and will try to update the scripts accordingly. I assume different versions will have different issues as well maybe.

**Reply by pcaversaccio** [reply]
Sharing some information here re `foundryup` since they recently moved to a new [organisation](https://github.com/foundry-rs/foundry): if anything broke, you might need to update your `foundryup`: `curl -L https://foundry.paradigm.xyz | sh`

**Reply by defido** [reply]
Also, do the first install in the main terminal, not in VSCODE, as I spent 15 minutes finding out :D
